### DISCUSSION TITLE
ADC Sample Speed

### DISCUSSION THREAD

**Discussion by NuclearPhoenixx** [earlephilhower/arduino-pico#1070] [Q&A] [answered]
Hello there again,

in the datasheet of the RP2040 it is stated that the ADC has a sample rate of 500 ksps. The ADC is much slower, though, if we use the standard `analogRead()` function that is implemented via the Pico SDK. Is the Pico SDK just much slower than the best possible sample rate, i.e. is there just so much programming overhead that it cannot reach the 500 ksps? Or can you get close to this via any of the Pico SDK functions? Maybe using the ADC FIFO would work faster? I'm not really sure, sorry if I'm just ignorant here.

https://github.com/earlephilhower/arduino-pico/blob/895ffced8e90372b6ec6e1052df87ad2800faa81/cores/rp2040/wiring_analog.cpp#L125-L142

**Comment by earlephilhower** [comment, marked as answer]
Using a simple test

````
void setup() {
  Serial.begin();
  delay(5000);
  uint32_t start = millis();
  int sum = analogRead(A0);
  for (int i = 0; i < 1000000; i++) {
    sum += analogRead(A0);
  }
  uint32_t end = millis();
  Serial.printf("ADC reads per second: %0.0f\n", 1000000.0 / ((end - start) * 0.001));
}

void loop() {
}
````

With the current core I get
````
ADC reads per second: 184298
````

Not 500k, but not completely atrocious.

With the simple optimizations in #1072 I get about 20% speed improvement:
````
ADC reads per second: 226347
````

The remaining difference is down to the SDK/manual reading of the ADC inputs which we have no control over.

**Reply by NuclearPhoenixx** [reply]
Thanks a lot for your testing and the explanation as always. You're doing a really great job, also thank you for https://github.com/earlephilhower/arduino-pico/pull/1072!

**Reply by earlephilhower** [reply]
@Phoenix1747 can you give #1072 a quick run?  You've obviously got an app using the ADC so it would be a good sanity check.

**Reply by NuclearPhoenixx** [reply]
@earlephilhower Yup, looks fine for me. It's also a bit faster indeed. Nice!

**Reply by earlephilhower** [reply]
Thx for the update, merged for the next release!
